File: cli/src/schema.rs
use anyhow::{anyhow, Context, Result};
use jsonschema::JSONSchema;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::fs;
use std::path::Path;

use crate::models::OakEntry;

/// Schema validator for Oak entries
pub struct SchemaValidator {
    schema: JSONSchema,
    schema_value: Value,
    enumerations: HashMap<String, Vec<String>>,
}

impl SchemaValidator {
    /// Load the schema from a file
    pub fn from_file<P: AsRef<Path>>(path: P) -> Result<Self> {
        let schema_content = fs::read_to_string(path.as_ref())
            .context("Failed to read schema file")?;

        let schema_value: Value = serde_json::from_str(&schema_content)
            .context("Failed to parse schema JSON")?;

        let schema = JSONSchema::compile(&schema_value)
            .map_err(|e| anyhow!("Failed to compile schema: {}", e))?;

        // Extract enumerations from the schema
        let enumerations = schema_value
            .get("enumerations")
            .and_then(|e| e.as_object())
            .map(|obj| {
                obj.iter()
                    .filter_map(|(k, v)| {
                        v.as_array().map(|arr| {
                            (
                                k.clone(),
                                arr.iter()
                                    .filter_map(|v| v.as_str().map(String::from))
                                    .collect(),
                            )
                        })
                    })
                    .collect()
            })
            .unwrap_or_default();

        Ok(Self {
            schema,
            schema_value,
            enumerations,
        })
    }

    /// Validate an oak entry
    pub fn validate(&self, entry: &OakEntry) -> Result<()> {
        // Convert to JSON for validation
        let json_value = serde_json::to_value(entry)
            .context("Failed to serialize entry")?;

        // Validate against schema
        if let Err(errors) = self.schema.validate(&json_value) {
            let error_messages: Vec<String> = errors
                .map(|e| format!("  - {}", e))
                .collect();

            return Err(anyhow!(
                "Validation failed:\n{}",
                error_messages.join("\n")
            ));
        }

        // Validate enumeration values
        self.validate_enumerations(entry)?;

        Ok(())
    }

    /// Validate that enumerated field values are in the allowed list
    fn validate_enumerations(&self, entry: &OakEntry) -> Result<()> {
        let mut errors = Vec::new();

        // Helper to check field values
        let mut check_field = |field_name: &str, values: &[crate::models::DataPoint]| {
            if let Some(allowed) = self.enumerations.get(field_name) {
                for dp in values {
                    if !allowed.contains(&dp.value) {
                        errors.push(format!(
                            "Invalid value '{}' for field '{}'. Allowed values: {}",
                            dp.value,
                            field_name,
                            allowed.join(", ")
                        ));
                    }
                }
            }
        };

        check_field("leaf_color", &entry.leaf_color);
        check_field("bud_shape", &entry.bud_shape);
        check_field("leaf_shape", &entry.leaf_shape);
        check_field("bark_texture", &entry.bark_texture);

        if !errors.is_empty() {
            return Err(anyhow!(
                "Enumeration validation failed:\n  - {}",
                errors.join("\n  - ")
            ));
        }

        Ok(())
    }

    /// Add a new enumeration value to a field
    pub fn add_enumeration_value(&mut self, field: &str, value: &str) -> Result<()> {
        if !self.enumerations.contains_key(field) {
            return Err(anyhow!("Field '{}' does not have enumeration validation", field));
        }

        let values = self.enumerations.get_mut(field).unwrap();
        if values.contains(&value.to_string()) {
            return Err(anyhow!(
                "Value '{}' already exists for field '{}'",
                value,
                field
            ));
        }

        values.push(value.to_string());
        values.sort();

        Ok(())
    }

    /// Save the schema back to file (used after adding enumeration values)
    pub fn save_to_file<P: AsRef<Path>>(&self, path: P) -> Result<()> {
        // Update the enumerations in the schema value
        let mut schema = self.schema_value.clone();
        if let Some(obj) = schema.as_object_mut() {
            let enum_obj: HashMap<String, Value> = self
                .enumerations
                .iter()
                .map(|(k, v)| {
                    let values: Vec<Value> = v.iter().map(|s| json!(s)).collect();
                    (k.clone(), json!(values))
                })
                .collect();

            obj.insert("enumerations".to_string(), json!(enum_obj));
        }

        let json_string = serde_json::to_string_pretty(&schema)
            .context("Failed to serialize schema")?;

        fs::write(path, json_string)
            .context("Failed to write schema file")?;

        Ok(())
    }

    /// Get the list of allowed values for a field
    pub fn get_allowed_values(&self, field: &str) -> Option<&[String]> {
        self.enumerations.get(field).map(|v| v.as_slice())
    }

    /// Get all enumerated fields
    pub fn get_enumerated_fields(&self) -> Vec<&str> {
        self.enumerations.keys().map(|s| s.as_str()).collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::DataPoint;

    #[test]
    fn test_validate_valid_entry() {
        let validator = SchemaValidator::from_file("schema/oak_schema.json").unwrap();

        let entry = OakEntry {
            scientific_name: "Quercus robur".to_string(),
            leaf_shape: vec![DataPoint {
                value: "lobed".to_string(),
                source_id: "src1".to_string(),
                page_number: None,
            }],
            ..OakEntry::new("Quercus robur".to_string())
        };

        assert!(validator.validate(&entry).is_ok());
    }

    #[test]
    fn test_validate_invalid_enum() {
        let validator = SchemaValidator::from_file("schema/oak_schema.json").unwrap();

        let entry = OakEntry {
            scientific_name: "Quercus robur".to_string(),
            leaf_shape: vec![DataPoint {
                value: "square".to_string(), // Invalid value
                source_id: "src1".to_string(),
                page_number: None,
            }],
            ..OakEntry::new("Quercus robur".to_string())
        };

        assert!(validator.validate(&entry).is_err());
    }
}

File: cli/internal/schema/schema.go
package schema

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/jeff/oaks/cli/internal/models"
	"github.com/santhosh-tekuri/jsonschema/v5"
)

// Validator handles JSON schema validation for Oak entries
type Validator struct {
	schema       *jsonschema.Schema
	enumerations map[string][]string
	schemaPath   string
}

// schemaWithEnums represents the schema file structure
type schemaWithEnums struct {
	Enumerations map[string][]string `json:"enumerations"`
}

// FromFile creates a new Validator from a schema file
func FromFile(schemaPath string) (*Validator, error) {
	data, err := os.ReadFile(schemaPath)
	if err != nil {
		return nil, fmt.Errorf("failed to read schema file: %w", err)
	}

	// Parse enumerations from schema
	var schemaData schemaWithEnums
	if err := json.Unmarshal(data, &schemaData); err != nil {
		return nil, fmt.Errorf("failed to parse schema: %w", err)
	}

	// Compile JSON Schema
	compiler := jsonschema.NewCompiler()
	if err := compiler.AddResource(schemaPath, strings.NewReader(string(data))); err != nil {
		return nil, fmt.Errorf("failed to add schema resource: %w", err)
	}

	schema, err := compiler.Compile(schemaPath)
	if err != nil {
		return nil, fmt.Errorf("failed to compile schema: %w", err)
	}

	return &Validator{
		schema:       schema,
		enumerations: schemaData.Enumerations,
		schemaPath:   schemaPath,
	}, nil
}

// ValidateOakEntry validates an OakEntry against the schema
func (v *Validator) ValidateOakEntry(entry *models.OakEntry) error {
	// First, validate against JSON Schema
	data, err := json.Marshal(entry)
	if err != nil {
		return fmt.Errorf("failed to marshal entry: %w", err)
	}

	var obj interface{}
	if err := json.Unmarshal(data, &obj); err != nil {
		return fmt.Errorf("failed to unmarshal entry: %w", err)
	}

	if err := v.schema.Validate(obj); err != nil {
		return fmt.Errorf("schema validation failed: %w", err)
	}

	// Validate enumerations
	if err := v.validateEnumerations(entry); err != nil {
		return err
	}

	return nil
}

// validateEnumerations checks field values against allowed enumerations
func (v *Validator) validateEnumerations(entry *models.OakEntry) error {
	var errors []string

	validateField := func(fieldName string, dataPoints []models.DataPoint) {
		allowedValues, hasEnum := v.enumerations[fieldName]
		if !hasEnum {
			return
		}

		allowed := make(map[string]bool)
		for _, val := range allowedValues {
			allowed[val] = true
		}

		for _, dp := range dataPoints {
			if !allowed[dp.Value] {
				errors = append(errors, fmt.Sprintf(
					"invalid value '%s' for field '%s'. Allowed values: %s",
					dp.Value, fieldName, strings.Join(allowedValues, ", "),
				))
			}
		}
	}

	validateField("leaf_shape", entry.LeafShape)
	validateField("leaf_color", entry.LeafColor)
	validateField("bud_shape", entry.BudShape)
	validateField("bark_texture", entry.BarkTexture)

	if len(errors) > 0 {
		return fmt.Errorf("enumeration validation failed:\n  - %s", strings.Join(errors, "\n  - "))
	}

	return nil
}

// AddEnumValue adds a new permitted value to a field's enumeration
func (v *Validator) AddEnumValue(field, value string) error {
	// Check if field has enumerations
	existing, ok := v.enumerations[field]
	if !ok {
		return fmt.Errorf("field '%s' does not have enumeration constraints", field)
	}

	// Check if value already exists
	for _, val := range existing {
		if val == value {
			return fmt.Errorf("value '%s' already exists for field '%s'", value, field)
		}
	}

	// Read the schema file
	data, err := os.ReadFile(v.schemaPath)
	if err != nil {
		return fmt.Errorf("failed to read schema file: %w", err)
	}

	var schemaMap map[string]interface{}
	if err := json.Unmarshal(data, &schemaMap); err != nil {
		return fmt.Errorf("failed to parse schema: %w", err)
	}

	// Update the enumerations
	enums, ok := schemaMap["enumerations"].(map[string]interface{})
	if !ok {
		return fmt.Errorf("invalid schema: missing enumerations")
	}

	fieldEnum, ok := enums[field].([]interface{})
	if !ok {
		return fmt.Errorf("invalid schema: field '%s' enumerations not found", field)
	}

	fieldEnum = append(fieldEnum, value)
	enums[field] = fieldEnum
	schemaMap["enumerations"] = enums

	// Write back the schema
	output, err := json.MarshalIndent(schemaMap, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal schema: %w", err)
	}

	if err := os.WriteFile(v.schemaPath, output, 0644); err != nil {
		return fmt.Errorf("failed to write schema file: %w", err)
	}

	// Update in-memory enumerations
	v.enumerations[field] = append(v.enumerations[field], value)

	return nil
}

// GetEnumValues returns the allowed values for a field
func (v *Validator) GetEnumValues(field string) ([]string, bool) {
	vals, ok := v.enumerations[field]
	return vals, ok
}

// GetAllEnumFields returns all fields that have enumeration constraints
func (v *Validator) GetAllEnumFields() []string {
	fields := make([]string, 0, len(v.enumerations))
	for field := range v.enumerations {
		fields = append(fields, field)
	}
	return fields
}
